File: .UNSAFE-ALL-BINDINGS-DELETED-FINAL-20260125/bindings/php/src/Proven/SafeUnit.php
<?php
// SPDX-License-Identifier: PMPL-1.0
// SPDX-FileCopyrightText: 2025 Hyperpolymath

declare(strict_types=1);

namespace Proven;

/**
 * Length units.
 */
enum LengthUnit: string
{
    case Meter = 'm';
    case Kilometer = 'km';
    case Centimeter = 'cm';
    case Millimeter = 'mm';
    case Mile = 'mi';
    case Yard = 'yd';
    case Foot = 'ft';
    case Inch = 'in';
    case NauticalMile = 'nmi';

    /**
     * Get conversion factor to meters.
     */
    public function toMeters(): float
    {
        return match ($this) {
            self::Meter => 1.0,
            self::Kilometer => 1000.0,
            self::Centimeter => 0.01,
            self::Millimeter => 0.001,
            self::Mile => 1609.344,
            self::Yard => 0.9144,
            self::Foot => 0.3048,
            self::Inch => 0.0254,
            self::NauticalMile => 1852.0,
        };
    }
}

/**
 * Mass units.
 */
enum MassUnit: string
{
    case Kilogram = 'kg';
    case Gram = 'g';
    case Milligram = 'mg';
    case Tonne = 't';
    case Pound = 'lb';
    case Ounce = 'oz';
    case Stone = 'st';

    /**
     * Get conversion factor to kilograms.
     */
    public function toKilograms(): float
    {
        return match ($this) {
            self::Kilogram => 1.0,
            self::Gram => 0.001,
            self::Milligram => 0.000001,
            self::Tonne => 1000.0,
            self::Pound => 0.45359237,
            self::Ounce => 0.028349523125,
            self::Stone => 6.35029318,
        };
    }
}

/**
 * Temperature units.
 */
enum TemperatureUnit: string
{
    case Celsius = 'C';
    case Fahrenheit = 'F';
    case Kelvin = 'K';
}

/**
 * Time units.
 */
enum TimeUnit: string
{
    case Second = 's';
    case Millisecond = 'ms';
    case Microsecond = 'us';
    case Nanosecond = 'ns';
    case Minute = 'min';
    case Hour = 'h';
    case Day = 'd';
    case Week = 'w';

    /**
     * Get conversion factor to seconds.
     */
    public function toSeconds(): float
    {
        return match ($this) {
            self::Second => 1.0,
            self::Millisecond => 0.001,
            self::Microsecond => 0.000001,
            self::Nanosecond => 0.000000001,
            self::Minute => 60.0,
            self::Hour => 3600.0,
            self::Day => 86400.0,
            self::Week => 604800.0,
        };
    }
}

/**
 * Data size units.
 */
enum DataUnit: string
{
    case Byte = 'B';
    case Kilobyte = 'KB';
    case Megabyte = 'MB';
    case Gigabyte = 'GB';
    case Terabyte = 'TB';
    case Petabyte = 'PB';
    case Kibibyte = 'KiB';
    case Mebibyte = 'MiB';
    case Gibibyte = 'GiB';
    case Tebibyte = 'TiB';
    case Pebibyte = 'PiB';

    /**
     * Get conversion factor to bytes.
     */
    public function toBytes(): float
    {
        return match ($this) {
            self::Byte => 1.0,
            self::Kilobyte => 1000.0,
            self::Megabyte => 1000000.0,
            self::Gigabyte => 1000000000.0,
            self::Terabyte => 1000000000000.0,
            self::Petabyte => 1000000000000000.0,
            self::Kibibyte => 1024.0,
            self::Mebibyte => 1048576.0,
            self::Gibibyte => 1073741824.0,
            self::Tebibyte => 1099511627776.0,
            self::Pebibyte => 1125899906842624.0,
        };
    }
}

/**
 * Safe unit conversion operations.
 */
class SafeUnit
{
    /**
     * Convert a length value between units.
     */
    public static function convertLength(float $value, LengthUnit $from, LengthUnit $to): float
    {
        $inMeters = $value * $from->toMeters();
        return $inMeters / $to->toMeters();
    }

    /**
     * Convert a mass value between units.
     */
    public static function convertMass(float $value, MassUnit $from, MassUnit $to): float
    {
        $inKg = $value * $from->toKilograms();
        return $inKg / $to->toKilograms();
    }

    /**
     * Convert a temperature value between units.
     */
    public static function convertTemperature(float $value, TemperatureUnit $from, TemperatureUnit $to): float
    {
        // Convert to Celsius first
        $celsius = match ($from) {
            TemperatureUnit::Celsius => $value,
            TemperatureUnit::Fahrenheit => ($value - 32) * 5 / 9,
            TemperatureUnit::Kelvin => $value - 273.15,
        };

        // Convert from Celsius to target
        return match ($to) {
            TemperatureUnit::Celsius => $celsius,
            TemperatureUnit::Fahrenheit => $celsius * 9 / 5 + 32,
            TemperatureUnit::Kelvin => $celsius + 273.15,
        };
    }

    /**
     * Convert a time value between units.
     */
    public static function convertTime(float $value, TimeUnit $from, TimeUnit $to): float
    {
        $inSeconds = $value * $from->toSeconds();
        return $inSeconds / $to->toSeconds();
    }

    /**
     * Convert a data size value between units.
     */
    public static function convertData(float $value, DataUnit $from, DataUnit $to): float
    {
        $inBytes = $value * $from->toBytes();
        return $inBytes / $to->toBytes();
    }

    /**
     * Format a byte count as human-readable string (binary units).
     */
    public static function formatBytes(float $bytes, int $precision = 2): string
    {
        $units = ['B', 'KiB', 'MiB', 'GiB', 'TiB', 'PiB'];
        $absBytes = abs($bytes);
        $sign = $bytes < 0 ? '-' : '';

        if ($absBytes < 1) {
            return $sign . '0 B';
        }

        $exp = min((int)floor(log($absBytes, 1024)), count($units) - 1);
        $value = $absBytes / pow(1024, $exp);

        return $sign . number_format($value, $precision) . ' ' . $units[$exp];
    }

    /**
     * Format a byte count as human-readable string (SI units).
     */
    public static function formatBytesSI(float $bytes, int $precision = 2): string
    {
        $units = ['B', 'KB', 'MB', 'GB', 'TB', 'PB'];
        $absBytes = abs($bytes);
        $sign = $bytes < 0 ? '-' : '';

        if ($absBytes < 1) {
            return $sign . '0 B';
        }

        $exp = min((int)floor(log($absBytes, 1000)), count($units) - 1);
        $value = $absBytes / pow(1000, $exp);

        return $sign . number_format($value, $precision) . ' ' . $units[$exp];
    }

    /**
     * Format a duration in seconds as human-readable string.
     */
    public static function formatDuration(float $seconds, bool $short = false): string
    {
        $seconds = abs($seconds);

        if ($seconds < 0.001) {
            $value = $seconds * 1000000;
            return number_format($value, 2) . ($short ? 'us' : ' microseconds');
        }
        if ($seconds < 1) {
            $value = $seconds * 1000;
            return number_format($value, 2) . ($short ? 'ms' : ' milliseconds');
        }
        if ($seconds < 60) {
            return number_format($seconds, 2) . ($short ? 's' : ' seconds');
        }
        if ($seconds < 3600) {
            $minutes = $seconds / 60;
            return number_format($minutes, 2) . ($short ? 'min' : ' minutes');
        }
        if ($seconds < 86400) {
            $hours = $seconds / 3600;
            return number_format($hours, 2) . ($short ? 'h' : ' hours');
        }

        $days = $seconds / 86400;
        return number_format($days, 2) . ($short ? 'd' : ' days');
    }

    /**
     * Parse a duration string (e.g., "1h30m", "2d", "500ms").
     *
     * @return Result<float> Duration in seconds
     */
    public static function parseDuration(string $input): Result
    {
        $input = trim(strtolower($input));
        if ($input === '') {
            return Result::err(ProvenError::emptyInput());
        }

        $total = 0.0;
        $pattern = '/(\d+(?:\.\d+)?)\s*(d|h|m|min|s|ms|us|ns)/';

        if (preg_match_all($pattern, $input, $matches, PREG_SET_ORDER) === 0) {
            // Try parsing as plain number (assumed seconds)
            if (is_numeric($input)) {
                return Result::ok((float)$input);
            }
            return Result::err(ProvenError::parseError('Invalid duration format'));
        }

        foreach ($matches as $match) {
            $value = (float)$match[1];
            $unit = $match[2];

            $multiplier = match ($unit) {
                'd' => 86400.0,
                'h' => 3600.0,
                'm', 'min' => 60.0,
                's' => 1.0,
                'ms' => 0.001,
                'us' => 0.000001,
                'ns' => 0.000000001,
                default => null,
            };

            if ($multiplier === null) {
                return Result::err(ProvenError::parseError("Unknown unit: $unit"));
            }

            $total += $value * $multiplier;
        }

        return Result::ok($total);
    }

    /**
     * Parse a byte size string (e.g., "1GB", "500MiB").
     *
     * @return Result<float> Size in bytes
     */
    public static function parseByteSize(string $input): Result
    {
        $input = trim($input);
        if ($input === '') {
            return Result::err(ProvenError::emptyInput());
        }

        $pattern = '/^(\d+(?:\.\d+)?)\s*(B|KB|MB|GB|TB|PB|KiB|MiB|GiB|TiB|PiB)?$/i';
        if (!preg_match($pattern, $input, $matches)) {
            return Result::err(ProvenError::parseError('Invalid byte size format'));
        }

        $value = (float)$matches[1];
        $unit = strtoupper($matches[2] ?? 'B');

        $multiplier = match ($unit) {
            'B' => 1.0,
            'KB' => 1000.0,
            'MB' => 1000000.0,
            'GB' => 1000000000.0,
            'TB' => 1000000000000.0,
            'PB' => 1000000000000000.0,
            'KIB' => 1024.0,
            'MIB' => 1048576.0,
            'GIB' => 1073741824.0,
            'TIB' => 1099511627776.0,
            'PIB' => 1125899906842624.0,
            default => null,
        };

        if ($multiplier === null) {
            return Result::err(ProvenError::parseError("Unknown unit: $unit"));
        }

        return Result::ok($value * $multiplier);
    }
}

File: .UNSAFE-ALL-BINDINGS-DELETED-FINAL-20260125/bindings/scala/src/main/scala/proven/SafeUnit.scala
// SPDX-License-Identifier: PMPL-1.0

package proven

/**
 * Unit categories for type safety.
 */
enum UnitCategory:
  case Length
  case Mass
  case Time
  case Temperature
  case DataSize
  case Speed
  case Area
  case Volume

/**
 * Quantity with unit for type-safe operations.
 */
case class Quantity[U <: UnitCategory](value: Double, unit: String, category: U):

  /**
   * Add two quantities (must be same category).
   */
  def +(other: Quantity[U]): Option[Quantity[U]] =
    if category != other.category then None
    else
      // Convert other to this unit
      SafeUnit.convert(other.value, other.unit, unit).map { converted =>
        Quantity(value + converted, unit, category)
      }

  /**
   * Subtract two quantities (must be same category).
   */
  def -(other: Quantity[U]): Option[Quantity[U]] =
    if category != other.category then None
    else
      SafeUnit.convert(other.value, other.unit, unit).map { converted =>
        Quantity(value - converted, unit, category)
      }

  /**
   * Multiply by scalar.
   */
  def *(scalar: Double): Quantity[U] = Quantity(value * scalar, unit, category)

  /**
   * Divide by scalar.
   */
  def /(scalar: Double): Option[Quantity[U]] =
    if scalar.abs < 1e-10 then None
    else Some(Quantity(value / scalar, unit, category))

  /**
   * Convert to another unit.
   */
  def to(targetUnit: String): Option[Quantity[U]] =
    SafeUnit.convert(value, unit, targetUnit).map(Quantity(_, targetUnit, category))

  override def toString: String = s"$value $unit"

/**
 * Safe unit conversion operations.
 */
object SafeUnit:

  // Conversion factors to base units
  private val lengthFactors: Map[String, Double] = Map(
    "m" -> 1.0,
    "km" -> 1000.0,
    "cm" -> 0.01,
    "mm" -> 0.001,
    "mi" -> 1609.344,
    "yd" -> 0.9144,
    "ft" -> 0.3048,
    "in" -> 0.0254,
    "nm" -> 1852.0,  // nautical mile
    "um" -> 1e-6,    // micrometer
    "nm" -> 1e-9     // nanometer (overrides nautical mile - use "nmi" for nautical)
  )

  private val massFactors: Map[String, Double] = Map(
    "kg" -> 1.0,
    "g" -> 0.001,
    "mg" -> 1e-6,
    "lb" -> 0.453592,
    "oz" -> 0.0283495,
    "t" -> 1000.0,
    "st" -> 6.35029  // stone
  )

  private val timeFactors: Map[String, Double] = Map(
    "s" -> 1.0,
    "ms" -> 0.001,
    "us" -> 1e-6,
    "ns" -> 1e-9,
    "min" -> 60.0,
    "h" -> 3600.0,
    "d" -> 86400.0,
    "wk" -> 604800.0
  )

  private val dataSizeFactors: Map[String, Double] = Map(
    "B" -> 1.0,
    "KB" -> 1024.0,
    "MB" -> 1024.0 * 1024,
    "GB" -> 1024.0 * 1024 * 1024,
    "TB" -> 1024.0 * 1024 * 1024 * 1024,
    "PB" -> 1024.0 * 1024 * 1024 * 1024 * 1024,
    "KiB" -> 1024.0,
    "MiB" -> 1024.0 * 1024,
    "GiB" -> 1024.0 * 1024 * 1024,
    "TiB" -> 1024.0 * 1024 * 1024 * 1024,
    "kB" -> 1000.0,
    "mB" -> 1000.0 * 1000,
    "gB" -> 1000.0 * 1000 * 1000
  )

  private val areaFactors: Map[String, Double] = Map(
    "m2" -> 1.0,
    "km2" -> 1e6,
    "cm2" -> 1e-4,
    "mm2" -> 1e-6,
    "ha" -> 10000.0,
    "acre" -> 4046.86,
    "sqft" -> 0.092903,
    "sqmi" -> 2.59e6
  )

  private val volumeFactors: Map[String, Double] = Map(
    "L" -> 1.0,
    "mL" -> 0.001,
    "m3" -> 1000.0,
    "cm3" -> 0.001,
    "gal" -> 3.78541,
    "qt" -> 0.946353,
    "pt" -> 0.473176,
    "floz" -> 0.0295735
  )

  /**
   * Convert between units.
   */
  def convert(value: Double, fromUnit: String, toUnit: String): Option[Double] =
    if fromUnit == toUnit then return Some(value)

    // Try each category
    convertWithFactors(value, fromUnit, toUnit, lengthFactors)
      .orElse(convertWithFactors(value, fromUnit, toUnit, massFactors))
      .orElse(convertWithFactors(value, fromUnit, toUnit, timeFactors))
      .orElse(convertWithFactors(value, fromUnit, toUnit, dataSizeFactors))
      .orElse(convertWithFactors(value, fromUnit, toUnit, areaFactors))
      .orElse(convertWithFactors(value, fromUnit, toUnit, volumeFactors))
      .orElse(convertTemperature(value, fromUnit, toUnit))

  private def convertWithFactors(
      value: Double,
      fromUnit: String,
      toUnit: String,
      factors: Map[String, Double]
  ): Option[Double] =
    for
      fromFactor <- factors.get(fromUnit)
      toFactor <- factors.get(toUnit)
    yield value * fromFactor / toFactor

  private def convertTemperature(
      value: Double,
      fromUnit: String,
      toUnit: String
  ): Option[Double] =
    // Convert to Kelvin first, then to target
    val kelvin: Option[Double] = fromUnit match
      case "K" => Some(value)
      case "C" | "degC" => Some(value + 273.15)
      case "F" | "degF" => Some((value - 32) * 5 / 9 + 273.15)
      case "R" => Some(value * 5 / 9)  // Rankine
      case _ => None

    kelvin.flatMap { k =>
      toUnit match
        case "K" => Some(k)
        case "C" | "degC" => Some(k - 273.15)
        case "F" | "degF" => Some((k - 273.15) * 9 / 5 + 32)
        case "R" => Some(k * 9 / 5)
        case _ => None
    }

  /**
   * Create a length quantity.
   */
  def length(value: Double, unit: String): Quantity[UnitCategory.Length.type] =
    Quantity(value, unit, UnitCategory.Length)

  /**
   * Create a mass quantity.
   */
  def mass(value: Double, unit: String): Quantity[UnitCategory.Mass.type] =
    Quantity(value, unit, UnitCategory.Mass)

  /**
   * Create a time quantity.
   */
  def time(value: Double, unit: String): Quantity[UnitCategory.Time.type] =
    Quantity(value, unit, UnitCategory.Time)

  /**
   * Create a data size quantity.
   */
  def dataSize(value: Double, unit: String): Quantity[UnitCategory.DataSize.type] =
    Quantity(value, unit, UnitCategory.DataSize)

  // Convenience conversions
  def metersToFeet(meters: Double): Double = meters / 0.3048
  def feetToMeters(feet: Double): Double = feet * 0.3048
  def milesToKm(miles: Double): Double = miles * 1.609344
  def kmToMiles(km: Double): Double = km / 1.609344
  def kgToPounds(kg: Double): Double = kg / 0.453592
  def poundsToKg(pounds: Double): Double = pounds * 0.453592
  def celsiusToFahrenheit(c: Double): Double = c * 9 / 5 + 32
  def fahrenheitToCelsius(f: Double): Double = (f - 32) * 5 / 9
  def litersToGallons(liters: Double): Double = liters / 3.78541
  def gallonsToLiters(gallons: Double): Double = gallons * 3.78541

  /**
   * Format bytes as human-readable string.
   */
  def formatBytes(bytes: Long): String =
    if bytes < 1024 then s"$bytes B"
    else if bytes < 1024 * 1024 then f"${bytes / 1024.0}%.1f KB"
    else if bytes < 1024 * 1024 * 1024 then f"${bytes / (1024.0 * 1024)}%.1f MB"
    else if bytes < 1024L * 1024 * 1024 * 1024 then f"${bytes / (1024.0 * 1024 * 1024)}%.1f GB"
    else f"${bytes / (1024.0 * 1024 * 1024 * 1024)}%.1f TB"

  /**
   * Format duration as human-readable string.
   */
  def formatDuration(seconds: Long): String =
    if seconds < 60 then s"${seconds}s"
    else if seconds < 3600 then s"${seconds / 60}m ${seconds % 60}s"
    else if seconds < 86400 then s"${seconds / 3600}h ${(seconds % 3600) / 60}m"
    else s"${seconds / 86400}d ${(seconds % 86400) / 3600}h"

  /**
   * Parse a value with unit (e.g., "10 kg", "5.5m").
   */
  def parse(s: String): Option[(Double, String)] =
    val pattern = """^(-?\d+\.?\d*)\s*([a-zA-Z]+\d*)$""".r
    s.trim match
      case pattern(value, unit) => Some((value.toDouble, unit))
      case _ => None
